quic/s2n-quic-core/src/recovery/bbr.rs

```diff
@@ -175,3 +207,116 @@ impl CongestionController for BbrCongestionController {
         todo!()
     }
 }
+
+impl BbrCongestionController {
+    /// The bandwidth-delay product
+    ///
+    /// Based on the current estimate of maximum sending bandwidth and minimum RTT
+    fn bdp(&self) -> u64 {
+        self.bdp_multiple(self.data_rate_model.bw(), Ratio::one())
+    }
+
+    /// Calculates a bandwidth-delay product using the supplied `Bandwidth` and `gain`
+    fn bdp_multiple(&self, bw: Bandwidth, gain: Ratio<u64>) -> u64 {
+        if let Some(min_rtt) = self.data_volume_model.min_rtt() {
+            (gain * (bw * min_rtt)).to_integer()
+        } else {
+            Self::initial_window(self.max_datagram_size).into()
+        }
+    }
+
+    /// How much data do we want in flight
+    ///
+    /// Based on the estimated BDP, unless congestion reduced the cwnd
+    fn target_inflight(&self) -> u32 {
+        self.bdp().min(self.cwnd as u64) as u32
+    }
+
+    /// The estimate of the volume of in-flight data required to fully utilize the bottleneck
+    /// bandwidth available to the flow, based on the BDP estimate (BBR.bdp), the aggregation
+    /// estimate (BBR.extra_acked), the offload budget (BBR.offload_budget), and BBRMinPipeCwnd.
+    #[allow(dead_code)] // TODO: Remove when used
+    fn max_inflight(&self) -> u64 {
+        let cwnd_gain = Ratio::one(); // TODO: get cwnd_gain from State
+        let bdp = self.bdp_multiple(self.data_rate_model.bw(), cwnd_gain);
+        let inflight = bdp + self.data_volume_model.extra_acked();
+        self.quantization_budget(inflight)
+    }
+
+    /// Inflight based on min RTT and the estimated bottleneck bandwidth
+    fn inflight(&self, bw: Bandwidth, gain: Ratio<u64>) -> u32 {
+        let inflight = self.bdp_multiple(bw, gain);
+        self.quantization_budget(inflight)
+            .try_into()
+            .unwrap_or(u32::MAX)
+    }
+
+    /// The volume of data that tries to leave free headroom in the bottleneck buffer or link for
+    /// other flows, for fairness convergence and lower RTTs and loss
+    fn inflight_with_headroom(&self) -> u32 {
+        if self.data_volume_model.inflight_hi() == u64::MAX {
+            return u32::MAX;
+        }
+
+        let headroom = max(
+            1,
+            (HEADROOM * self.data_volume_model.inflight_hi()).to_integer(),
+        );
+        max(
+            self.data_volume_model.inflight_hi() - headroom,
+            self.minimum_window() as u64,
+        )
+        .try_into()
+        .unwrap_or(u32::MAX) // TODO: change type
+    }
+
+    /// Calculates the quantization budget
+    fn quantization_budget(&self, inflight: u64) -> u64 {
```

Comment:
My understanding its basically the amount of data we should have in flight to fully utilize the path, given that both the sender and receiver may have aggregation mechanisms or delayed acknowledgements that would cause some of the data to be buffered on either end.